core/lib/dal/src/events_web3_dal.rs

```diff
@@ -13,16 +14,61 @@ pub struct EventsWeb3Dal<'a, 'c> {
     pub(crate) storage: &'a mut StorageProcessor<'c>,
 }
 
+pub fn filter_logs_by_api_eth_transfer_events(
+    mut db_logs: Vec<StorageWeb3Log>,
+    api_eth_transfer_events: ApiEthTransferEvents,
+) -> Vec<StorageWeb3Log> {
+    if api_eth_transfer_events == ApiEthTransferEvents::Disabled {
+        db_logs.retain(|log| {
+            log.address != L2_ETH_TOKEN_ADDRESS.as_bytes()
+                || log.topic1 != TRANSFER_EVENT_TOPIC.as_bytes()
+        });
+
+        if let Some(first_log) = db_logs.first().cloned() {
+            let mut new_event_index_in_block = 0;
+            let mut new_event_index_in_tx = 0;
+            let mut current_block = first_log.miniblock_number;
+            let mut current_tx_hash = first_log.tx_hash.clone();
+
+            for db_log in &mut db_logs {
+                if db_log.miniblock_number != current_block {
+                    current_block = db_log.miniblock_number;
+                    new_event_index_in_block = 0;
+                }
+                if db_log.tx_hash != current_tx_hash {
+                    current_tx_hash = db_log.tx_hash.clone();
+                    new_event_index_in_tx = 0;
+                }
+                db_log.event_index_in_block = new_event_index_in_block;
+                db_log.event_index_in_tx = new_event_index_in_tx;
```

Comment:
Maybe a stupid question, but how will it work when not all logs from block are queried? I.e. if block has 2 events: first with topic1=A, second with topic1=B, both events are not ETH transfer. If someone sends a request with filter.topic1=B, won't the response will have the only log with event_index_in_block=0, while it should be 1?

Reply:
That's actually a good question, I didn't really think about it, and yes, it won't work properly. I will think how to fix that behaviour, but first question that comes to my mind - do you know how bad could it be if we try to filter topics/addresses on Rust side?

Reply:
hmm, I don't think it's possible with the current approach, we don't limit block range at all, so worst case we need to filter millions of events

Reply:
One way to make it work without too much changes:
query ETH transfer events regardless provided filter and then count how many ETH transfers were before each event within block/tx and subtract this number from indices